onnxruntime/core/graph/graph.cc

```diff
@@ -4015,7 +4015,7 @@ Status Graph::InlineFunction(Node& callnode) {
   // create a uniq_identifier to append to every node name and intermediate input\outputs
   // to make sure there are no unintended duplicates
   std::stringstream ss;
-  ss << "_" << static_cast<const void*>(&callnode) << "_";
+  ss << "_" << static_cast<const void*>(&callnode) << "_" << callnode.Name() << "_";
```

Comment:
If the previous generated name used the address of the calling node, wouldn't the name of the calling node also be the same in both instances?

Reply:
Evidently it is not. In general, there is no guarantee that for the entire runtime a unique object exists at a specific address.
This being the in-line function handler it could be using/reusing same graph only with different names. This being nested function nodes make it possible to not uniquify only by using the address. Taking name without the address also causes duplicate name error. The original reported error is caused for the Shape node.

The following graph uses nested functions.
![image](https://github.com/microsoft/onnxruntime/assets/26722914/21859f92-03a3-48ea-9db0-882db1f991ae)
Both Conv2d(L__self__conv1)_0 and Conv2d(L__self__conv2)_3 use _aten_convolution_onnx_1 function
![image](https://github.com/microsoft/onnxruntime/assets/26722914/737bf5fd-58fc-4f9e-b8cc-12c49dc60170)  
![image](https://github.com/microsoft/onnxruntime/assets/26722914/028b9630-6b14-4412-9308-95d1ab0faab0)

The following is the graph for _aten_convolution_onnx
![image](https://github.com/microsoft/onnxruntime/assets/26722914/ec2b0862-d398-4f0f-a27d-f63a693138dd)

Reply:
When you get the duplicate call at the same address what is the optype/name in each call, and is the same of a different Graph instance (e.g. main or a subgraph)? 

Trying to understand how this can happen given InlineFunction is called from GraphPartitioner and none of the Nodes being inlined change during that iteration (i.e. the addresses for call_node in inlining for a Graph instance do not change during inlining).

Reply:
It looks something like the "_0x56346bee65b0_n0" as specified in the issue description. The address followed by "_n0".
"n0" is the name of the shape node in the graph.

Reply:
It's not clear to me how two different nodes that are being inlined have the same address. 

Does inlining one node insert a new node that also needs to be inlined? 
- That could possibly explain it if the memory used for the Node instance that was inlined and freed, was somehow re-used for a new Node being created by inlining later on 
- But if not, I would have expected all the nodes that can be inlined would exist during model loading and therefore could not have the same address. 

Also not clear how the Shape node is being passed to InlineFunction as `callnode` for its name to be in that string given Shape is not defined as a function in the ONNX spec. AFAIK we only call Graph::InlineFunction here: 

https://github.com/microsoft/onnxruntime/blob/648bedf91a9e390fadbbabdbbb9d5239f5a9d201/onnxruntime/core/framework/graph_partitioner.cc#L520-L528

and that is only for nodes that are function based (e.g. LeakyRelu is defined in the [ONNX spec](https://github.com/onnx/onnx/blob/main/docs/Operators.md) as a function)

https://github.com/microsoft/onnxruntime/blob/648bedf91a9e390fadbbabdbbb9d5239f5a9d201/onnxruntime/core/graph/graph.cc#L569-L578

Without understanding how the duplicate address is occurring it's hard to say if adding the name from the second call node is the correct fix. e.g. what if something else is off and we shouldn't be calling Graph::InlineFunction with the node that has a duplicate address - like somehow we've created a Shape node with a function body and that shouldn't actually happen?

Reply:
Anyway for uniqueness we should depend on names not the addresses. The node names are traceable back to the source. If the input has same name for different nodes then it would be the user error. The model is hierarchical and one branch after the other are in-lined.  This explains why incidentally two nodes are processed with same address but with different names.

Reply:
Nodes don't have to have names according to the ONNX spec.

What explains the name of a Shape node being in callnode.Name()?

Reply:
Yes. That is why I kept the original address part.

Reply:
If the address is not guaranteed to be unique, what guarantees that both `callnode` values will have a name? Without that the new identifier isn't guaranteed to be unique is it, and if not we still have the original issue?

Reply:
In order to clarify how/why node address is being repeated, I added some debug output statements locally in the code to print, the list of nodes being in-lined, node addresses, name, op types and new nodes added. The function nodes are in-lined breadth first by the graph partitioner. 
After in-lining the function nodes are removed [here](https://github.com/microsoft/onnxruntime/blob/e9e6bedf37428e394bf24c2e706e734748e7374c/onnxruntime/core/graph/graph.cc#LL4122C13-L4122C13). New nodes are created at the same address during in-lining another function node.

Before in-lining aten_detach_2 function node exists at 0x2bd3e80. It is removed and released. A new node is created at the same address with name _0x2bd4440__aten_convolution_onnx_1. In the second pass the new node is also in-lined.
In this case two node one with op type Identity and the other with op type Shape end of having same name "_0x2bd3e80_n0". There may be other nodes with duplicate names.

graph_partitioner.cc:528  Begin nodes to inline:
torch.nn.modules.conv.Conv2d(L__self___conv1)_0
aten_sigmoid_1
aten_detach_2
torch.nn.modules.conv.Conv2d(L__self___conv2)_3
aten_sigmoid_4
aten_detach_5
graph_partitioner.cc:532  End nodes to inline.
graph.cc:4020  callnode = _0x2a3f700_  name = torch.nn.modules.conv.Conv2d(L__self___conv1)_0 op_type = torch.nn.modules.conv.Conv2d(L__self___conv1)
graph.cc:4061  Added node at = 0x2ca5c10  name = _0x2a3f700__aten_convolution_onnx_1 op_type = _aten_convolution_onnx
graph.cc:4020  callnode = _0x2bd38d0_  name = aten_sigmoid_1 op_type = aten_sigmoid
graph.cc:4061  Added node at = 0x2a3f700  name = _0x2bd38d0_n0 op_type = Sigmoid
graph.cc:4020  callnode = **_0x2bd3e80**_  name = aten_detach_2 op_type = aten_detach
graph.cc:4061  Added node at = 0x2bd38d0  name = **_0x2bd3e80_n0** op_type = Identity
graph.cc:4020  callnode = _0x2bd4440_  name = torch.nn.modules.conv.Conv2d(L__self___conv2)_3 op_type = torch.nn.modules.conv.Conv2d(L__self___conv2)
graph.cc:4061  Added node at = **0x2bd3e80**  name = _0x2bd4440__aten_convolution_onnx_1 op_type = _aten_convolution_onnx
graph.cc:4020  callnode = _0x2bd47a0_  name = aten_sigmoid_4 op_type = aten_sigmoid
graph.cc:4061  Added node at = 0x2bd4440  name = _0x2bd47a0_n0 op_type = Sigmoid
graph.cc:4020  callnode = _0x2bd4d70_  name = aten_detach_5 op_type = aten_detach
graph.cc:4061  Added node at = 0x2bd47a0  name = _0x2bd4d70_n0 op_type = Identity
graph_partitioner.cc:528  Begin nodes to inline:
_0x2a3f700__aten_convolution_onnx_1
_0x2bd4440__aten_convolution_onnx_1
graph_partitioner.cc:532  End nodes to inline.
graph.cc:4020  callnode = _0x2ca5c10_  name = _0x2a3f700__aten_convolution_onnx_1 op_type = _aten_convolution_onnx
graph.cc:4061  Added node at = 0x2bd4d70  name = _0x2ca5c10_n0 op_type = Shape
graph.cc:4061  Added node at = 0x2caf730  name = _0x2ca5c10_n1 op_type = Size
graph.cc:4061  Added node at = 0x2caf990  name = _0x2ca5c10_n2 op_type = Shape
graph.cc:4061  Added node at = 0x2cafbf0  name = _0x2ca5c10_n3 op_type = Size
graph.cc:4061  Added node at = 0x2cafe50  name = _0x2ca5c10_n4 op_type = Equal
graph.cc:4061  Added node at = 0x2cb00b0  name = _0x2ca5c10_n5 op_type = Not
graph.cc:4061  Added node at = 0x2cb0580  name = _0x2ca5c10_n6 op_type = If
graph.cc:4061  Added node at = 0x2cb1e10  name = _0x2ca5c10_n7 op_type = If
graph.cc:4061  Added node at = 0x2cb6df0  name = _0x2ca5c10_n8 op_type = If
graph.cc:4020  callnode = _**0x2bd3e80**_  name = _0x2bd4440__aten_convolution_onnx_1 op_type = _aten_convolution_onnx
graph.cc:4061  Added node at = 0x2ca5c10  name = **_0x2bd3e80_n0** op_type = Shape
graph.cc:4061  Added node at = 0x2cc0410  name = _0x2bd3e80_n1 op_type = Size
graph.cc:4061  Added node at = 0x2cc15d0  name = _0x2bd3e80_n2 op_type = Shape
graph.cc:4061  Added node at = 0x2cc1830  name = _0x2bd3e80_n3 op_type = Size
graph.cc:4061  Added node at = 0x2cc1a90  name = _0x2bd3e80_n4 op_type = Equal
graph.cc:4061  Added node at = 0x2cc1cf0  name = _0x2bd3e80_n5 op_type = Not
graph.cc:4061  Added node at = 0x2cc2150  name = _0x2bd3e80_n6 op_type = If
graph.cc:4061  Added node at = 0x2cbccc0  name = _0x2bd3e80_n7 op_type = If
graph.cc:4061  Added node at = 0x2cc8080  name = _0x2bd3e80_n8 op_type = If
Traceback (most recent call last):
  File "test.py", line 2, in <module>
    onnxruntime.InferenceSession("nested_function.onnx")
  File "/home/sajandhy/miniconda3/envs/py38/lib/python3.8/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 383, in __init__
    self._create_inference_session(providers, provider_options, disabled_optimizers)
  File "/home/sajandhy/miniconda3/envs/py38/lib/python3.8/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 435, in _create_inference_session
    sess.initialize_session(providers, provider_options, disabled_optimizers)
onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : This is an invalid model. Error: two nodes with same node name (_**0x2bd3e80**_n0).

Reply:
I think adding the name of the callnode (function node being in-lined) as a suffix to the nodes is sufficient to solve the problem as function_utils::specialize is called the rename the nodes using the uniq_identifier. Because the function_node names are expected to be unique with in their scope, multiple pass of in-lining will not result in duplicate names. One alternative is to exhaustively check the graph to make sure that no node exists in the graph with the same name before adding a new name. This could be expensive.

Reply:
So it sounds like neither the address nor the callnode.Name() have any uniqueness guarantees. If I have nested functions and the function definition does not name any nodes I have the potential for address re-use as well as multiple callnode instances with no name don't I? 

Can we maybe use graph::GenerateNodeName to generate the prefix? That helper is used by all the optimizers to create valid node names so using here would be good for consistency. Whilst we would be using it to create a node name prefix rather than a node name I don't think that matters. If that works, a `base_name` of maybe 'Inline' + callnode.OpType() may be helpful.

Reply:
I have modified code to use GenerateNodeName to uniquify new node names.